### DISCUSSION TITLE
How to combine tar_force() with dynamic branching?

### DISCUSSION THREAD

**Discussion by MxNl** [ropensci/tarchetypes#28] [Help]
I need to combine tar_force() with dynamic branching. Currently, I don't see a pattern argument in tar_force(). How can I do this? Thanks a lot as usual for you great support!

As a workaround I thought of using a helper target like
```{r}
tar_force(
   target_name, 
   Sys.time(),
   force = !dbExistsTable(...)
),
tar_target(
   create_table_in_db
)
```
to create a table in a database if it doesnt already exist.

**Comment by wlandau** [comment]
I am avoiding built-in dynamic branching in `tar_force()`, `tar_change()`, etc. because there is potential for confusion. Would you elaborate on your use case?

**Reply by gofford** [reply]
Related problem, so posting here rather then in a new thread:

Use-case: I have many models in a dynamic-within-static branching DAG. For each mapped `value` (static branch), there is a dynamic branch of models that are trained with the prediction subsequently combined with `tar_combine()`. Desired workflow is that the models train once a week, whilst the pipeline uses the weekly-trained model to make daily predictions.

In this scenario, having `tar_force()` or `tar_change()` (or even `tar_cue_age()`) act at a per-branch level in a dynamic pattern means that the individual models can be re-trained discretely if needed (such that if 1 model fails for any reason, it's not necessary to re-train all 16).

**Reply by wlandau** [reply]
Cues and cue-based functionality are defined statically and are not compatible with dynamic branching. But for your situation, there is a relatively straightforward workaround: look in the metadata, find the targets whose time stamps are old, and call `tar_invalidate()` on all of them. You can run this just before every `tar_make()`:

```r
library(targets)
na_false <- function(x) {
  x[is.na(x)] <- FALSE
  x
}
meta <- tar_meta()
threshold <- as.difftime(1, units = "weeks") # 1 week
old_targets <- meta$name[na_false(Sys.time() - meta$time) > threshold]
tar_invalidate(all_of(old_targets)) # all_of() is necessary here.
```

**Reply by gofford** [reply]
I hadn't considered that, and it's actually pretty clever. Combining the above in a `invalidate_old(name, age)` function with `tar_cue(mode = 'never')` on the model step seems to meet the requirement. Thanks @wlandau!

**Reply by wlandau** [reply]
Glad it works for you. I just added a new `tar_older()` function to `targets`, which should simplify this to something like `tar_invalidate(all_of(tar_older(Sys.time - as.difftime(1, units = "weeks"))))` .

**Reply by gofford** [reply]
> Cues and cue-based functionality are defined statically and are not compatible with dynamic branching. 

Does this mean that `tar_cue_age()` can be expected to work on a per-branch basis if the branches are generated statically by `tar_map()`? I have a scenario where this _isn't_ happening, but want to validate the expected behaviour before I raise a separate discussion to troubleshoot.

**Reply by wlandau** [reply]
Yes, all cues are compatible with static branching.

**Reply by gofford** [reply]
So if I have something like this, where `model_index` is pulled in from a map:

```
tar_target_raw(
   name = "model_fitted",
      command = substitute(
         expr = lightgbm_forecaster(split, model_index, target, nthread, archive_path),
         env = list(
            split = quote(model_data$split),
                model_index = quote(model_index),
                target = config$model$target,
                nthread = config$nthread,
                archive_path = config$targets$folders$models
             )
         ),
         cue = tar_cue_age_raw(
            name = "model_fitted",
            age = as.difftime(7, units = 'days')
         )
     )
```

expectation is that each of the `model_fitted_{0,1....16}` will be invalidated based on their own individual age, rather than across all of the different mapped elements?

**Reply by wlandau** [reply]
Hmm... I actually take that back `tar_map()` modifies the `command` and `pattern` part of targets, but not `tar_cue_age()`. `tar_cue_age()` needs to be unique to each static branch because it needs the target name. But `tar_eval()` should work: https://books.ropensci.org/targets/static.html#metaprogramming
